src/tox/tox_env/python/api.py

```diff
@@ -132,9 +132,40 @@ def extract_base_python(env_name: str) -> str | None:
         candidates: list[str] = []
         for factor in env_name.split("-"):
             spec = PythonSpec.from_string_spec(factor)
-            impl = spec.implementation or "python"
-            if impl.lower() in INTERPRETER_SHORT_NAMES and env_name is not None and spec.path is None:
+
+            # The regex used to parse a Python spec in PythonSpec.from_string_spec is extraordinarily loose so we do our
+            # own sanity checks. In short, if the user does not specify an implementation then a major version will
+            # be present and we insist that it must look like a known Python major version number. If they did specify
+            # an implementation then a major version might not be present, but we will verify it if it is (for
+            # consistency).
+            if (
+                factor != "py"
+                and (  # py is a special case where version info can be unset
+                    (spec.implementation and spec.major) or not spec.implementation
+                )
+                and spec.major not in (2, 3)
+            ):
+                continue
+
+            impl = spec.implementation
+
+            # TODO: Drop in a future version
+            # PythonSpec.from_string_spec set implementation to None for py and python
+            if impl is None and not factor.startswith("py") and not factor.startswith("python"):
+                logging.warning(
+                    "identified the %s factor in the %s testenv as a Python spec but the interpreter implementation "
+                    "was not specified. This will be a ignored in future versions of tox. Please prefix this factor "
+                    "with an interpreter short name (one of: %s)",
+                    factor,
+                    env_name,
+                    ", ".join(list(INTERPRETER_SHORT_NAMES) + ["py"]),
+                )
```

Comment:
Remove this warning please. `tox r -e 3.8` I think should be valid and we should default to `py` prefix in that case.